### PR TITLE
feat(wallet): starting Address Index loading, picking, exporting

### PR DESCRIPTION
**Pull request by destrys**
## Description
Caravan is now aware of the starting address index. 
When loading a wallet config caravan will jump ahead to the starting address index.
When creating a new wallet, a new Starting Address Index Picker is available.

## Does this PR introduce a breaking change?

* [ ] Yes
* [x] No

## Does this PR fixes open issues?

* [x] Yes
* [ ] No
#149

### CODE REVIEW

**Comment by waldenraines** [comment]
Screenshot of selector:

gscreenshot_2020-07-02-141530

File: src/components/Wallet/ConfirmWallet.jsx [hunk 1]
```diff
@@ -25,6 +27,11 @@ class WalletConfirmation extends React.Component {
           </TableHead>
           <TableBody>{this.renderConfirmationInfo()}</TableBody>
         </Table>
+        {startingAddressIndex > 0 && (
+          <Typography variant="subtitle1">
+            Attention: Starting Address Index set to {startingAddressIndex}
+          </Typography>
```
Comment (waldenraines) [inline src/components/Wallet/ConfirmWallet.jsx:33]: To me this looks odd:

gscreenshot_2020-07-02-145856

I think we should use `<Alert severity="info">` or `<Alert severity="warning">` instead.  @bucko13 thoughts?

gscreenshot_2020-07-02-145828



See also https://material-ui.com/components/alert/

Comment (destrys) [reply src/components/Wallet/ConfirmWallet.jsx:33]: Agreed, I think I like Alert Info.

File: src/components/StartingAddressIndexPicker.jsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      setStartingAddressIndex(parseInt(index, 10));
+    }
+    this.setState({
+      startingAddressIndexField: index,
+      startingAddressIndexError: error,
+    });
+  };
+
+  handleCustomIndexChange = (event) => {
+    const customIndex = event.target.value === "true";
```
[the remark this answers is not in the document]
Comment (bucko13) [reply src/components/StartingAddressIndexPicker.jsx:61]: Sweet sweet javascript.

File: src/components/Wallet/ConfirmWallet.jsx (hunk 1 above)
Comment (bucko13) [reply src/components/Wallet/ConfirmWallet.jsx:33]: So the idea would be info alert if it's a custom start index and nothing if it's default or for all circumstances? 

For a future update, I kind of like the idea of using a chip collection for the other items in the config (network, index, address type) could be a nice way to be able to display this type of info. https://material-ui.com/components/chips/

**Comment by bucko13** [comment]
> Additionally I think the index selector would be better displayed by changing the card heading of "Address Type" to "Address" and having it in that card (perhaps optionally hidden by a "show advanced" toggle).

I like this idea! Maybe "Address configuration"?

File: src/components/Wallet/ConfirmWallet.jsx (hunk 1 above)
Comment (waldenraines) [reply src/components/Wallet/ConfirmWallet.jsx:33]: > So the idea would be info alert if it's a custom start index and nothing if it's default or for all circumstances?

Yeah, exactly.

> For a future update, I kind of like the idea of using a chip collection for the other items in the config (network, index, address type) could be a nice way to be able to display this type of info. https://material-ui.com/components/chips/

Agreed.

File: src/components/StartingAddressIndexPicker.jsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  RadioGroup,
+  TextField,
+  Tooltip,
+} from "@material-ui/core";
+
+import { wrappedActions } from "../actions/utils";
+import { SET_STARTING_ADDRESS_INDEX } from "../actions/settingsActions";
+
+class StartingAddressIndexPicker extends React.Component {
+  // TODO: Add this validation to unchained-bitcoin
```
[the remark this answers is not in the document]
Comment (bucko13) [reply src/components/StartingAddressIndexPicker.jsx:22]: Any reason not to add this for this release? Should be a quick process.

[the remark this answers is not in the document]
Comment (bucko13) [reply src/components/StartingAddressIndexPicker.jsx:22]: I also believe some of this is in unchained-bitcoin already: https://github.com/unchained-capital/unchained-bitcoin/blob/master/src/paths.js#L146

[the remark this answers is not in the document]
Comment (destrys) [reply src/components/StartingAddressIndexPicker.jsx:22]: I didn't want to deal with bumping the unchained-bitcoin version, this needs to get in asap.
I was using the `validateBIP32PathSegment function`, but it wasn't correctly returning an error for a large integer.

**Comment by destrys** [comment]
I don't want to mess with the other settings pickers. While I agree the starting address index and the address type could be co-located, I don't want this to get held up by UX debate.

**Comment by waldenraines** [comment]
> I don't want to mess with the other settings pickers. While I agree the starting address index and the address type could be co-located, I don't want this to get held up by UX debate.

I'm honestly fine with that considering that we are already aware that we want to rework this.

File: src/components/Wallet/ConfirmWallet.jsx [hunk 4]
```diff
@@ -25,6 +27,11 @@ class WalletConfirmation extends React.Component {
           </TableHead>
           <TableBody>{this.renderConfirmationInfo()}</TableBody>
         </Table>
+        {startingAddressIndex > 0 && (
+          <Alert severity="info">
+            Setting Address Index set to {startingAddressIndex}
+          </Alert>
+        )}
```
Comment (waldenraines) [inline src/components/Wallet/ConfirmWallet.jsx:34]: What do you think about this?

```diff
diff --git a/src/components/Wallet/ConfirmWallet.jsx b/src/components/Wallet/ConfirmWallet.jsx
index 53fd28d..93e905a 100644
--- a/src/components/Wallet/ConfirmWallet.jsx
+++ b/src/components/Wallet/ConfirmWallet.jsx
@@ -17,6 +17,12 @@ class WalletConfirmation extends React.Component {
     const { startingAddressIndex } = this.props;
     return (
       <Box>
+        {startingAddressIndex > 0 && (
+          <Alert severity="info">
+            Setting Address Index set to {startingAddressIndex}
+          </Alert>
+        )}
+
         <Table>
           <TableHead>
             <TableRow>
@@ -27,11 +33,6 @@ class WalletConfirmation extends React.Component {
           </TableHead>
           <TableBody>{this.renderConfirmationInfo()}</TableBody>
         </Table>
-        {startingAddressIndex > 0 && (
-          <Alert severity="info">
-            Setting Address Index set to {startingAddressIndex}
-          </Alert>
-        )}
       </Box>
     );
   };
```

Basically moving the alert up to the top of the box.

gscreenshot_2020-07-06-074432

File: src/components/StartingAddressIndexPicker.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (waldenraines) [reply src/components/StartingAddressIndexPicker.jsx:22]: Okay, well I think we should add an issue to record this so we don't forget to add this to unchained-bitcoin and use it from there.

[the remark this answers is not in the document]
Comment (bucko13) [reply src/components/StartingAddressIndexPicker.jsx:22]: Yeah, I was just thinking about this today. I’d argue for at least doing it in parallel. I don’t see this delaying anything by more than an hour or two (at least not the version bumping) and it gives us the benefit of avoiding tech debt.

[the remark this answers is not in the document]
Comment (bucko13) [reply src/components/StartingAddressIndexPicker.jsx:22]: So if it becomes a blocker we can get this out as is. If it’s done and published in time, we can remove this and bump the local version for release.

File: src/components/Wallet/ConfirmWallet.jsx (hunk 4 above)
Comment (destrys) [reply src/components/Wallet/ConfirmWallet.jsx:34]: That's a typo, should be 'Starting Address Index'.
I agree on placement, making this change.

File: src/components/StartingAddressIndexPicker.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (destrys) [reply src/components/StartingAddressIndexPicker.jsx]: done

File: src/components/Wallet/ConfirmWallet.jsx (hunk 1 above)
Comment (destrys) [reply src/components/Wallet/ConfirmWallet.jsx]: done.

**Comment by bucko13** [comment]
Yessir! Looks good to me.

**Pull request merged**